node/src/node/builder.rs

```diff
@@ -0,0 +1,353 @@
+    pub fn sampling_window(self, dur: Duration) -> Self {
+        NodeBuilder {
+            sampling_window: Some(dur),
+            ..self
+        }
+    }
+
+    /// Set pruning delay.
+    ///
+    /// Pruning delay how much time the pruner should wait after syncing window in
+    /// order to prune the block.
+    ///
+    /// **Default if [`InMemoryStore`] is used:** 60 seconds.\
+    /// **Default:** 1 hour.\
+    /// **Minimum:** 60 seconds.
+    pub fn pruning_delay(self, dur: Duration) -> Self {
+        NodeBuilder {
+            pruning_delay: Some(dur),
+            ..self
+        }
+    }
+
+    fn build_config(self) -> Result<NodeConfig<B, S>, NodeBuilderError> {
+        let network = self.network.expect("todo");
+
+        let bootnodes = if self.bootnodes.is_empty() {
+            network.canonical_bootnodes().collect()
+        } else {
+            self.bootnodes
+        };
+
+        if bootnodes.is_empty() && self.listen.is_empty() {
+            // It is a valid scenario for user to create a node without any bootnodes
+            // and listening addresses. However it may not be what they wanted. Because
+            // of that we display a warning.
+            warn!("Node has empty bootnodes and listening addresses. It will never connect to another peer.");
+        }
+
+        // `Node` is memory hungry when in-memory stores are used and the user most
+        // probably will never know that they need to have a smaller syncing window
+        // to avoid it. In order to make `lumina` more user friendly we decided to
+        // have a different default when our in-memory stores are used.
+        //
+        // If user implements their own in-memory stores then they are responsible
+        // to set the syncing window to something smaller than `DEFAULT_SYNCING_WINDOW`.
+        let in_memory_store_used = TypeId::of::<S>() == TypeId::of::<InMemoryStore>();
+        let in_memory_blockstore_used = TypeId::of::<B>() == TypeId::of::<InMemoryBlockstore>();
+
+        let syncing_window = if let Some(dur) = self.syncing_window {
+            dur
+        } else if in_memory_store_used {
+            MIN_SYNCING_WINDOW
+        } else {
+            DEFAULT_SYNCING_WINDOW
+        };
+
+        let sampling_window = if let Some(dur) = self.syncing_window {
+            dur
+        } else if in_memory_blockstore_used {
+            MIN_SAMPLING_WINDOW
+        } else {
+            DEFAULT_SAMPLING_WINDOW
+        };
+
+        let pruning_delay = if let Some(dur) = self.pruning_delay {
+            dur
+        } else if in_memory_store_used {
+            MIN_PRUNING_DELAY
+        } else {
+            DEFAULT_PRUNING_DELAY
+        };
+
+        if syncing_window < MIN_SYNCING_WINDOW {
+            return Err(NodeBuilderError::VerySmallSynincWindow(syncing_window));
+        }
+
+        // Truncate sampling window if needed.
+        let sampling_window = sampling_window.min(syncing_window);
```

Comment:
I was debating if we should truncate sampling window or return an error if it is bigger than syncing window. What do you think?